### ISSUE TITLE
Incorrect type detection from TypeVar with default value

### ISSUE THREAD

**Issue by last-partizan** [microsoft/pyright#3614]
When using default value with TypeVar (like this `Type[T] = int`), it resolves to `Type[T@..] | Type[int]` instead of just `Type[T@..]`.

Here's example:

```python
from typing import TypeVar, Type

T = TypeVar("T", int, float)


def convert_number(v, t: Type[T] = int) -> T:
    reveal_type(t)  # Type of "t" is "Type[T@convert_number] | Type[int]"
    return t(v)  # - error: Expression of type "T@convert_number | int"
                 # cannot be assigned to return type "T@convert_number"


int_val_default = convert_number(1)
reveal_type(int_val_default)  # Type of "int_val_default" is "Unknown"

int_val = convert_number(1, int)
reveal_type(int_val)  #  Type of "int_val" is "Unknown"

float_val = convert_number(1, float)
reveal_type(float_val)  # Type of "float_val" is "float"
```

**Expected behavior**
I expect this to type check without errors, and int_val to be detected as `int`

It works fine with mypy:

```
> mypy return_type.py
return_type.py:7: note: Revealed type is "Type[builtins.int]"
return_type.py:7: note: Revealed type is "Type[builtins.float]"
return_type.py:12: note: Revealed type is "builtins.int"
return_type.py:15: note: Revealed type is "builtins.int"
return_type.py:18: note: Revealed type is "builtins.float"
Success: no issues found in 1 source file
```

**Additional context**
- pyright 1.1.255

Is this a bug in pyright or am i doing something wrong?

**Comment by erictraut** [comment]
This behavior is intended. Providing a default argument for a parameter that has a generic type is fraught with problems from a type safety perspective. Other Python type checkers like mypy don't allow this combination. Pyright allows it, but it makes it safe by assuming that type of the parameter is a union of the generic type and the default argument type.

Here is how I would recommend implementing this:

```python
@overload
def convert_number(v: object) -> int: ...

@overload
def convert_number(v: object, t: type[T]) -> T: ...

def convert_number(v: object, t: type = int) -> Any:
    return t(v)
```
